src/foam/box/pipeline/PipelineManager.js

```diff
@@ -0,0 +1,258 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+foam.CLASS({
+  package: 'foam.box.pipeline',
+  name: 'PipelineManager',
+
+  documentation: `A manager for composing pipelines of foam.box.Runnable
+    instances. Pipelines may fork via multiple then()s. Manager's current
+    runnables may be bound together with bind(); this can be used to merge
+    pipelines. Pipelines may not contain cycles.
+
+    Note that each runnable is registered as a service exactly once, even if
+    a builder referring to pipeline stages is built multiple times.
+
+    E.g.,
+
+    var b = PipelineManager.create();
+    var shared = b.then(sharedRunnable);
+
+    shared.then(forkRunnable1);
+    shared.then(forkRunnable2);
+
+    var merge1Builder = PipelineManager.create().then(mergeRunnable1);
+    var merge2Builder = PipelineManager.create().then(mergeRunnable1);
+    merge1Builder.bind(shared);
+    merge2Builder.bind(shared);
+
+    var inputToMerge1Runnable = merge1Builder.build();
+    var inputToMerge2Runnable = merge2Builder.build();
+    // Yields input boxes for mergeRunnable1 and mergeRunnable2
+    // on pipeline:
+    //
+    // mergeRunnable1 --                       -- forkRunnable1
+    //                  >-- sharedRunnable -- <
+    // mergeRunnable2 --                       -- forkRunnable2`,
+
+  requires: [
+    'foam.box.Box',
+    'foam.box.BroadcastBox',
+    'foam.box.LogBox',
+    'foam.box.Message',
+    'foam.box.RPCMessage',
+    'foam.box.RPCReturnBox',
+    'foam.box.pipeline.PipelineNode'
+  ],
+
+  imports: [
+    'registry',
+    'defaultOutputBox? as ctxDefaultOutputBox'
+  ],
+
+  classes: [
+    {
+      name: 'RunnableRPCBox',
+      extends: 'foam.box.ProxyBox',
+
+      requires: [
+        'foam.box.Message',
+        'foam.box.RPCMessage'
+      ],
+
+      properties: [
+        {
+          class: 'FObjectProperty',
+          of: 'foam.box.Box',
+          documentation: `Error box for routing errors when sending to
+              runnable.`,
+          name: 'errorBox'
+        },
+      ],
+
+      methods: [
+        function send(inputMessage) {
+          return this.delegate && this.delegate.send(this.Message.create({
+            object: this.RPCMessage.create({
+              name: 'run',
+              args: [ inputMessage.object ],
+            }),
+            attributes: { errorBox: this.errorBox }
+          }));
+        }
+      ]
+    }
+  ],
+
+  properties: [
+    {
+      class: 'FObjectProperty',
+      of: 'foam.box.Box',
+      name: 'defaultOutputBox',
+      documentation: `Output box used for end-of-computation (when no
+          next-to-run delegates bound to this step in the pipeline).`,
+      factory: function() {
+        return this.ctxDefaultOutputBox || this.LogBox.create();
```

Comment:
A similar warning (`Access missing import: ctxDefaultOutputBox`) will be displayed here as well